File: programmers/level-2/17683. 방금그곡/solution.cc
#include <string>
#include <vector>
#include <iostream>

using namespace std;

vector<string> split_string(string& origin, char delimiter) {
    vector<string> splited;
    int prev = 0, curr = origin.find(delimiter);

    while (curr != string::npos) {
        string substring = origin.substr(prev, curr - prev);
        splited.push_back(substring);
        prev = curr + 1;
        curr = origin.find(',', prev);
    }
    splited.push_back(origin.substr(prev, curr - prev));
    return splited;
}

int calc_play_time(string& begin, string& end) {
    vector<string> vb = split_string(begin, ':');
    vector<string> ve = split_string(end, ':');
    return (stoi(ve[0]) * 60 + stoi(ve[1])) - (stoi(vb[0]) * 60 + stoi(vb[1]));
}

string replace_sharp(string& note) {
    string replaced = "";
    for (int i = 1; i <= note.size(); i++) {
        if (note[i] == '#') {
            replaced += tolower(note[i-1]);
            i++;
        }
        else {
            replaced += note[i-1];
        }
    }
    return replaced;
}

string solution(string m, vector<string> musicinfos) {
    string origin = replace_sharp(m);
    string res_title = "(None)";
    int res_size = 0;

    for (auto it = musicinfos.begin(); it != musicinfos.end(); it++) {
        vector<string> vec = split_string(*it, ',');
        string title = vec[2];
        string replaced = replace_sharp(vec[3]);
        string target = replaced;
        int playtime = calc_play_time(vec[0], vec[1]);

        if (playtime > replaced.length()) {
            for (int i = 0; i < playtime - replaced.length(); i++) {
                target += target[i];
            }
        }
        else {
            target = replaced.substr(0, playtime);
        }

        if (target.find(origin) != string::npos) {
            if (res_size < target.length()) {
                res_title = title;
                res_size = target.length();
            }
        }
    }
    return res_title;
}

int main() {
    cout << solution("ABCDEFG", {"12:00,12:14,HELLO,CDEFGAB", "13:00,13:05,WORLD,AB#CDE#F#"});
    cout << solution("ABC", {"12:00,12:14,HELLO,C#DEFGAB", "13:00,13:05,WORLD,ABCDEF"});
    return 0;
}

File: programmers/level-2/17683. 방금그곡/solution.js
/**
 * https://programmers.co.kr/learn/courses/30/lessons/17683
 * @param {string} m 
 * @param {string[]} musicinfos 
 * @returns {string}
 */
function solution(m, musicinfos) {
    const possibles = [];
    const origin = replaceSharp(m);

    for (let i = 0; i < musicinfos.length; i += 1) {
        const [begin, end, title, note] = musicinfos[i].split(',');
        const playtime = minute(end) - minute(begin);
        const replaced = replaceSharp(note);
        let target = replaced;

        if (playtime > replaced.length) {
            for (let k = 0; target.length < playtime; k += 1) {
                target += target[k];
            }
        }
        else {
            target = replaced.slice(0, playtime);
        }

        if (target.includes(origin)) {
            possibles.push([title, playtime, i]);
        }
    }

    possibles.sort((a, b) => cmp(a, b));
    return possibles.length ? possibles[0][0] : '(None)';
}

function minute(formattedTime) {
    let [hh, mm] = formattedTime.split(':').map(v => parseInt(v, 10));
    if (hh === 0) hh = 24;
    return (hh * 60) + mm;
}

function replaceSharp(target) {
    return target.replace(/(\D)#/g, (match) => match[0].toLowerCase());
}

function cmp(a, b) {
    if (a[1] > b[1]) return -1;
    else if (a[1] < b[1]) return 1;
    else {
        if (a[2] > b[2]) return 1;
        else if (a[2] < b[2]) return -1;
        else return 0;
    }
}

solution('ABCDEFG', ["12:00,12:14,HELLO,CDEFGAB", "13:00,13:05,WORLD,ABCDEF", "12:00,12:10,HELLO,C#B#ABC"])
// solution('CC#BCC#BCC#BCC#B', ["03:00,03:30,FOO,CC#B", "04:00,04:08,BAR,CC#BCC#BCC#B", "05:00,05:3,FOO3,CC#B"])
// solution('ABC', ["12:00,12:14,HELLO,C#DEFGAB", "13:00,13:05,WORLD,ABCDEF"])
// solution('ABC', ["12:00,12:4,HELLO,CAB", "13:00,13:04,WORLD,ABC", "13:00,13:4,FUCK,ABCA#"])